Source/Scene/parseKTX2.js

```diff
@@ -0,0 +1,225 @@
+    (dfd.colorModel === colorModelETC1S || dfd.colorModel === colorModelUASTC)
+  ) {
+    // Compressed, initialize transcoder module
+    transcodeCompressed(
+      data,
+      header,
+      supportedTargetFormats,
+      transcoderModule,
+      result
+    );
+  } else {
+    parseUncompressed(header, result);
+  }
+
+  return result;
+}
+
+// Parser for uncompressed
+function parseUncompressed(header, result) {
+  var internalFormat = PixelFormat.RGBA;
+  var datatype;
+  if (header.vkFormat === VulkanConstants.VK_FORMAT_R8G8B8A8_UNORM) {
+    datatype = PixelDatatype.UNSIGNED_BYTE;
+  } else if (
+    header.vkFormat === VulkanConstants.VK_FORMAT_R16G16B16A16_SFLOAT
+  ) {
+    datatype = PixelDatatype.HALF_FLOAT;
+  } else if (
+    header.vkFormat === VulkanConstants.VK_FORMAT_R32G32B32A32_SFLOAT
+  ) {
+    datatype = PixelDatatype.FLOAT;
+  }
+
+  for (var i = 0; i < header.levels.length; ++i) {
+    var level = (result[i] = {});
+    var levelBuffer = arraySlice(header.levels[i].levelData);
+
+    var width = header.pixelWidth >> i;
+    var height = header.pixelHeight >> i;
+
+    for (var j = 0; j < header.faceCount; ++j) {
+      var faceLength =
+        header.typeSize *
+        width *
+        height *
+        PixelFormat.componentsLength(internalFormat);
+
+      var faceView;
+      if (PixelDatatype.sizeInBytes(datatype) === 1) {
+        faceView = new Uint8Array(levelBuffer.buffer, faceLength * j);
+      } else if (PixelDatatype.sizeInBytes(datatype) === 2) {
+        faceView = new Uint16Array(levelBuffer.buffer, faceLength * j);
+      } else {
+        faceView = new Float32Array(levelBuffer.buffer, faceLength * j);
+      }
+
+      level[faceOrder[j]] = {
+        internalFormat: internalFormat,
+        datatype: datatype,
+        width: width,
+        height: height,
+        levelBuffer: faceView,
+      };
+    }
+  }
+}
+
+function transcodeCompressed(
+  data,
+  header,
+  supportedTargetFormats,
+  transcoderModule,
+  result
+) {
+  var dataBuffer = defined(data.buffer) ? data.buffer : data;
+  var ktx2File = new transcoderModule.KTX2File(new Uint8Array(dataBuffer));
+  var width = ktx2File.getWidth();
+  var height = ktx2File.getHeight();
+  // var layers = ktx2File.getLayers();
+  var levels = ktx2File.getLevels();
+  // var faces = ktx2File.getFaces();
+  var hasAlpha = ktx2File.getHasAlpha();
+
+  if (!width || !height || !levels) {
+    ktx2File.close();
+    ktx2File.delete();
+    throw new RuntimeError("Invalid .ktx2 file");
+  }
+
+  // Determine target format based on platform support
+  var internalFormat, transcoderFormat;
+  var BasisFormat = transcoderModule.transcoder_texture_format;
+  if (supportedTargetFormats.astc && hasAlpha) {
+    internalFormat = PixelFormat.RGBA_ASTC;
+    transcoderFormat = BasisFormat.cTFASTC_4x4_RGBA;
+  } else if (supportedTargetFormats.bc7) {
+    internalFormat = PixelFormat.RGBA_BC7;
+    transcoderFormat = BasisFormat.cTFBC7_RGBA;
```

Comment:
Couple questions about the precedence of GPU compressed texture formats

* What's the advantage of BC7 over DXT5? Is BC7 higher quality? Is it better for normal maps?
* This order makes more sense for UASTC, but should ETC1S transcode to ETC1/DXT1 by default? They are smaller in GPU memory than ASTC/BC7 and might be good enough quality.

Reply:
- BC7 is higher quality than the older BC formats; I found [this post](https://www.reedbeta.com/blog/understanding-bcn-texture-compression-formats/#bc1) comparing them:
![image](https://user-images.githubusercontent.com/31491650/122684259-6a330d00-d1d2-11eb-9e45-bfd8d647995e.png)
![image](https://user-images.githubusercontent.com/31491650/122684254-5edfe180-d1d2-11eb-8e72-b7ea0317e4c7.png)
Though based on their example of a full res texture maybe DXT5 is sufficient for CesiumJS?
![image](https://user-images.githubusercontent.com/31491650/122684299-a1092300-d1d2-11eb-885f-a46f8b2a05fb.png)
![image](https://user-images.githubusercontent.com/31491650/122684331-cb5ae080-d1d2-11eb-9470-df0c3097bfc1.png)
- According to [these comments](https://github.com/BinomialLLC/basis_universal/blob/c4c0db735a6a799702cd4e7309df64b4f7f59e39/webgl/ktx2_encode_test/index.html#L312) in the Binomial WebGL encoding example, the preferred format for ETC1S without alpha is ETC1. Not sure how they decided that but the format precedence in their code doesn't seem to force ETC1 for all ETC1S files either... I think the idea is that ETC1 does not support alpha, so it should not be used unless the ktx2 is in ETC1S + no alpha. Otherwise a format with an alpha channel is needed like BC7 (according to the above post and some other sources DXT1 only supports a 1-bit alpha channel?).